src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -367,7 +378,61 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
         container.pendingContext = context;
       }
 
-      scheduleTopLevelUpdate(current, element, callback);
+      scheduleTopLevelUpdate(current, element, false, callback);
+    },
+
+    // Like updateContainer, but blocks the root from committing. Returns an
+    // expiration time.
+    // TODO: Can this be unified with updateContainer? Or is it incompatible
+    // with the existing semantics of ReactDOM.render?
+    updateRoot(
+      element: ReactNodeList,
+      container: OpaqueRoot,
+      parentComponent: ?React$Component<any, any>,
+      callback: ?Function,
+    ): ExpirationTime {
+      // TODO: If this is a nested container, this won't be the root.
+      const current = container.current;
+
+      if (__DEV__) {
+        if (ReactFiberInstrumentation.debugTool) {
+          if (current.alternate === null) {
+            ReactFiberInstrumentation.debugTool.onMountContainer(container);
+          } else if (element === null) {
+            ReactFiberInstrumentation.debugTool.onUnmountContainer(container);
+          } else {
+            ReactFiberInstrumentation.debugTool.onUpdateContainer(container);
+          }
+        }
+      }
+
+      const context = getContextForSubtree(parentComponent);
+      if (container.context === null) {
+        container.context = context;
+      } else {
+        container.pendingContext = context;
+      }
+
+      return scheduleTopLevelUpdate(current, element, true, callback);
+    },
+
+    unblockRoot(root: OpaqueRoot, expirationTime: ExpirationTime) {
```

Comment:
Why do we need these to be separate? This is inherent complexity to support where as the current API only lets you do both at once.

I think we should try again to simply defer all commits until explicitly called since we have so many cases where that seems to matter. And the legacy mode is simply a wrapper that does that eagerly. In that case that's the normal mode so the concept of being blocked doesn't make sense and unblocking doesn't as well.